Pcap++/src/PcapFilter.cpp

```diff
@@ -297,60 +297,23 @@ void EtherTypeFilter::parseToString(std::string& result)
 	result = "ether proto " + stream.str();
 }
 
-AndFilter::AndFilter(std::vector<GeneralFilter*>& filters)
-{
-	for(std::vector<GeneralFilter*>::iterator it = filters.begin(); it != filters.end(); ++it)
-	{
-		m_FilterList.push_back(*it);
-	}
-}
-
-void AndFilter::setFilters(std::vector<GeneralFilter*>& filters)
-{
-	m_FilterList.clear();
-
-	for(std::vector<GeneralFilter*>::iterator it = filters.begin(); it != filters.end(); ++it)
-	{
-		m_FilterList.push_back(*it);
-	}
-}
+CompositeFilter::CompositeFilter(const std::vector<GeneralFilter*>& filters) : m_FilterList(filters) {}
 
-void AndFilter::parseToString(std::string& result)
+void CompositeFilter::removeFilter(GeneralFilter* filter)
 {
-	result.clear();
-	for(std::vector<GeneralFilter*>::iterator it = m_FilterList.begin(); it != m_FilterList.end(); ++it)
+	for(auto it = m_FilterList.cbegin(); it != m_FilterList.cend(); ++it)
 	{
-		std::string innerFilter;
-		(*it)->parseToString(innerFilter);
-		result += '(' + innerFilter + ')';
-		if (m_FilterList.back() != *it)
+		if (*it == filter)
 		{
-			result += " and ";
+			m_FilterList.erase(it);
+			break;
 		}
 	}
 }
 
-OrFilter::OrFilter(std::vector<GeneralFilter*>& filters)
+void CompositeFilter::setFilters(const std::vector<GeneralFilter*>& filters)
 {
-	for(std::vector<GeneralFilter*>::iterator it = filters.begin(); it != filters.end(); ++it)
-	{
-		m_FilterList.push_back(*it);
-	}
-}
-
-void OrFilter::parseToString(std::string& result)
-{
-	result.clear();
-	for(std::vector<GeneralFilter*>::iterator it = m_FilterList.begin(); it != m_FilterList.end(); ++it)
-	{
-		std::string innerFilter;
-		(*it)->parseToString(innerFilter);
-		result += '(' + innerFilter + ')';
-		if (m_FilterList.back() != *it)
-		{
-			result += " or ";
-		}
-	}
+	m_FilterList = filters;
```

Comment:
question, clearing the origin vector and adding elements v.s. copy assignment, which one performance is better?

Reply:
Since C++11 copy assign attempts to reuse memory block of the vector being assigned to, if the allocator is not propagated by the vector being copied from or both vectors use the same allocator. [Link to cpp reference](https://en.cppreference.com/w/cpp/container/vector/operator%3D#:~:text=If,wise%20copy%2Dassignment.)

The default allocator does not propagate on copy so most of the time the vector block will be reused if possible.
Provided no reallocations happen due to insufficient capacity, copy assignment ends up being slightly faster due to less function calls overhead.

Ran a quick benchmark on https://perfbench.com/ and got those results:
NB: Horrible formatting, I know. Might be better actually running the sample on the site as it gives a boxplot result too.

ID                                          Calls    Median       0ns      32ns     128ns     512ns       2us       8us      33us     131us     524us       2ms       8ms      34ms     134ms     537ms
copy_assign_initial_empty                    2000      33us         0         0         0         0         0         0      1996         0         0         0         4         0         0         0
clear\_&\_push_back_initial_empty              2000     131us         0         0         0         0         0         0         0      1938         0         0        62         0         0         0
clear\_&\_push_back_initial_has_data           2000      33us         0         0         0         0         0         0      1980         0         0         0        20         0         0         0
copy_assign_enough_capacity                  2000      33us         0         0         0         0         0         0      1993         0         0         0         7         0         0         0
copy_assign_require_capacity_expansion       2000     131us         0         0         0         0         0         0         0      1969         0         0        31         0         0         0

ID                                          Calls        Min        Avg        Max      StDev
copy_assign_initial_empty                    2000     41.816     68.928    13.05ms    581.028 
clear\_&\_push_back_initial_empty              2000    388.390    795.693    13.40ms     2.25ms 
clear\_&\_push_back_initial_has_data           2000    126.226    256.895    13.13ms     1.29ms 
copy_assign_enough_capacity                  2000     44.616     91.881    13.05ms    768.054 
copy_assign_require_capacity_expansion       2000    193.277    401.726    13.21ms     1.61ms 


Benchmark code here (requires to set timeout to 5secs):
https://gist.github.com/Dimi1010/3cdecaa4a0a94df29d1862add64ea9d5

Reply:
thanks for explaining, well I didn't see any drawback to have a clean code now :)

btw, next time, you can just have a screen shot than raw data.